feat: :sparkles: add "writing Python" guide section (#48)
seedcase-project/guidebook: python/code.qmd

user to understand what the code does.

### Type hints {#type-hints}

For keeping the input and output consistent and clear, it helps to write
[type hints](https://docs.python.org/3/library/typing.html). What are
type hints in Python? They are a way to formally declare the type of
object that is given or output by a function or method. For example, for
the function `check_heart_rate()`, you might write:

``` {.python filename="src/health_data_checks/check_heart_rate.py"}
def check_heart_rate(data: dict) -> bool:
    """Check if the heart rate is within a normal range."""
    # Code ...
    return boolean_output
```

Where `data: dict` tells Python that the input to the function must be a
[`dict`](https://realpython.com/python-dicts/) object and the `-> bool`
tells Python that the output of the function has to be a boolean value
(`True` or `False`). So if you were to look at the documentation for the
function or method, you will be able to read what object types the code
needs and what it outputs.

::: callout-note
Since you're using mypy when using a project made from
`template-python-package`, type hints are required. Otherwise mypy will

> give an error. Unlike other programming languages, Python does not
> enforce type hints when using the code as it is a [dynamically
> typed](https://en.wikipedia.org/wiki/Python_(programming_language)).

Reviewer: ```suggestion
give an error. Unlike some other programming languages, Python does not
enforce type hints when the code is run, as it is [dynamically
typed](https://en.wikipedia.org/wiki/Python_(programming_language)).
```

This means that you can use any type of object as input to the function,
even if it is not the type you declared in the type hint. While this is
a feature of Python, it does make it more difficult to write code that
is predictable and consistent.
:::

### Side effects {#side-effects}

> Side effects are when the code does something that isn't obvious, isn't
> clear from the name of the function or method, or if it does actions
> that don't result in a `return`.

Reviewer: ```suggestion
Side effects happen when the code does something that isn't obvious
from the name of the function or method, or when it performs actions
without returning a value.
```

Some side effects are intentional and can't be avoided, like writing to
a file or database, which are inherently side effects. But in general,

> you want to avoid side effects as that introduces unpredictability in

Reviewer: ```suggestion
you want to avoid side effects as they introduce unpredictability in
```

your code.

For example, if you use a function called `check_heart_rate()`, you
probably expect it to output a boolean value that tells you if the heart
rate is normal or not. However, if it also writes to a file or modifies
the data in some way, then it has a side effect. A side effect for
`check_heart_rate()` might look like:

``` {.python filename="src/health_data_checks/check_heart_rate.py"}
def check_heart_rate(data: dict) -> bool:
    """Check if the heart rate is within a normal range."""
    # Side effect: Writing

>     data.write_csv("heart_rate.csv")

Reviewer: I know this is pseudo code, but if we type hint `data` as a `dict`, then it feels weird using a non-existent method on it

Other side effects include modifying objects in the global scope, such
as modifying the input arguments. Using the same `check_heart_rate()`
function, a side effect that modifies the input might look like this:

``` {.python filename="src/health_data_checks/check_heart_rate.py"}
# Fake data
data = {
    "heart_rate": 30,
}

def check_heart_rate(data: dict) -> bool:
    """Check if the heart rate is within a normal range."""
    # Side effect: Modifying the input in the global scope.
    not_normal = data["heart_rate"] < 30 or data["heart_rate"] > 180
    data.update(check=not_normal)
    return not_normal

# Original data
print(data)
print(check_heart_rate(data))
# Modified data
print(data)
```

A function should never modify objects in the global scope or modify the
input arguments, as it makes it much harder to understand, predict, and
test.

### Internal code

Rarely does a public-facing function or method do everything within
itself. That would end up making the code very long and difficult to
read. Instead, you create smaller, well-defined internal functions,
methods, or classes that do very specific actions and use them to build
up the larger public-facing functionality.

> As already described above in the [where to write

Reviewer: ```suggestion
As already described above in [where to write
```

code](#where-to-write-code), when more than one public-facing function
or method uses the same internal functionality, keep that internal
functionality in the `internals/` folder.

How you write this internal functionality is basically no different from
writing public-facing functionality. The only difference is that you
should use an underscore (`_`) before the name of the object.

Then, once you've made the function or method, in order to "export" it
to make it easier to import, you open the `__init__.py` file in the
`internals/` folder and write it like:

``` {.python filename="src/health_data_check/internals/__init__.py"}
# Import the `is.py` script by using the relative
# `.` import syntax.
from .is import _is_data_object

# Use `__all__` to explicitly define what is exported
# from this module.
__all__ = [
    "_is_data_object",
]
```

> That way, when you import the internal functionality in other scripts,

Reviewer: ```suggestion
This way, when you import the internal functionality in other scripts,
```

you can use either:

``` python
from internals import _is_data_object

# Using the function.
_is_data_object(data)
```

Or:

``` python
import internals

# Using the function
internals._is_data_object(data)
```

We tend to prefer the first option, as it makes the code a bit cleaner

> and clearer about what exactly is being used in the script.

Reviewer: The second half of this sentence is losing me a bit

## Writing code documentation

Documentation is one of the most important parts of writing code.

> Without good documentation, code is near to unusable or at the least
> extremely difficult to use. Documenting at the code level is done with

Reviewer: ```suggestion
Without good documentation, code is nearly unusable or at least
extremely difficult to use. Documenting at the code level is done with
```

["docstrings"](https://peps.python.org/pep-0257/) in Python.

> At the least, you should write a docstring for every public-facing

Reviewer: ```suggestion
At the very least, you should write a docstring for every public-facing
```

    use `path: The path to the file.` rather than
    `path: the path to the file`.
-   Include example code in the Quarto code chunk format.
-   Write simple one-line docstrings for:
    -   Very small functions or methods (for example, \<5 lines)
    -   Functions or methods with \<2 arguments
    -   Functions or methods with simple and clear input and output
-   Write longer and more detailed docstrings for:
    -   Larger and/or more complicated internal functions, classes, or
        methods, such as those with non-standard input or output
    -   User-facing functions, classes, or methods

::: callout-tip
When using VS Code and adding docstrings, use the
[autoDocstring](https://marketplace.visualstudio.com/items?itemName=njpwerner.autodocstring)
VS Code extension to automatically generate a skeleton structure for the
docstrings.
:::

::: content-hidden
Need to use `{{python}}` in example below because Quarto seems to interpret that
as a code block. It renders fine though.
:::
```` {.python filename="src/health_data_checks/check_heart_rate.py"}
def check_heart_rate(data: dict) -> bool:
    """Check if the heart rate is within a normal range.

    Args:

>         data (dict): The data object containing the

Reviewer: Do we usually include the type here?

Guide that can be automatically checked and fixed. The tools we use and
recommend are [Ruff](https://decisions.seedcase-project.org/why-ruff/)
for general styling and [mypy](https://mypy.readthedocs.io/en/stable/)
for type checking.

To use Ruff and mypy, you need to be in a Python project. If you have
created a Python project with our
[`template-python-package`](https://github.com/seedcase-project/template-python-package),
then running Ruff can be done through our
[justfile](https://just.systems/man/en/). In your terminal, you can run
this:

``` {.bash filename="Terminal"}
just format-python
```

Which will run Ruff and format your Python code. Then, to check if your code
needs manual fixes using Ruff and mypy, you can check the code with:

``` {.bash filename="Terminal"}
just check-python
```

::: callout-tip
If you didn't use our template nor do you want to add Ruff to your
project's dependencies, you can use Ruff through
[uv](https://docs.astral.sh/uv/):

``` {.bash filename="Terminal"}
uvx ruff format
# Tries to fix the linted issues, a bit different from `format`
uvx ruff check --fix
uvx ruff check
uvx mypy .
```
:::

> ## Where to write code {#where-to-write-code}

Reviewer 2: Stumbled in the same way as Signe on the Where. Maybe 'How to structure your code' or something like that. In addition to the solution provided by Signe.

`from internals import FUNCTION` when needing to use the internal
functionality.

## How to write code

As in nearly all programming languages, how you write code is up to you
as the developer. The exact details depend heavily on the problem your
software is trying to solve, the design, the constraints, and the
requirements.

However, there are some good general practices to follow that will help
you write more effective and maintainable Python code.

### Avoiding "magic"

Code that is "magic" does something unexpected or
unpredictable. Usually this is because it has ["side
effects"](#side-effects) or conditionally changes its action in
fundamental ways that are not always obvious.

The way you avoid magic is by writing code that is simple, predictable,
and consistent. This means that writing code in a more [functional
programming](https://en.wikipedia.org/wiki/Functional_programming) style
tends to produce more maintainable and more predictable code. Strategies
for writing more functional code include:

-   Using clear inputs and one type of output for functions and methods.
    There is more explanation on this in the [type hints](#type-hints)
    section below.

> -   Avoiding [side effects](#side-effects) by only outputting the result of the function

Reviewer 2: Should this have a reference to the section on side effects in the same way as type hints above?

easier to create unit [tests](tests.qmd) for it and it is easier for the
user to understand what the code does.

### Type hints {#type-hints}

For keeping the input and output consistent and clear, it helps to write
[type hints](https://docs.python.org/3/library/typing.html). What are
type hints in Python? They are a way to formally declare the type of
object that is the input or output by a function or method. For example, for
the function `check_heart_rate()`, you might write:

``` {.python filename="src/health_data_checks/check_heart_rate.py"}
def check_heart_rate(data: dict[str, Any]) -> bool:
    """Check if the heart rate is within a normal range."""
    # Code ...
    return boolean_output
```

Where `data: dict[str, Any]` tells Python that the input to the function
must be a [`dict`](https://realpython.com/python-dicts/) object with
`str` or `Any` as internal data types and the `-> bool` tells Python
that the output of the function has to be a boolean value (`True` or
`False`). So if you look at the documentation for the function or
method, you will be able to read what object types the code needs and
what it outputs.

::: callout-note

> Since you're using mypy when using a project made from

Reviewer 2: Should this be re-phrased to say something like 'If you have based your project on our template, then you will be using mypy which requires type hints'?

Guide that can be automatically checked and fixed. The tools we use and
recommend are [Ruff](https://decisions.seedcase-project.org/why-ruff/)
for general styling and [mypy](https://mypy.readthedocs.io/en/stable/)
for type checking.

To use Ruff and mypy, you need to be in a Python project. If you have
created a Python project with our
[`template-python-package`](https://github.com/seedcase-project/template-python-package),
then running Ruff can be done through our
[justfile](https://just.systems/man/en/). In your terminal, running
this:

``` {.bash filename="Terminal"}
just format-python
```

Will run Ruff and format your Python code. Then, to check if your code
needs manual fixes using Ruff and mypy, you can check the code with:

``` {.bash filename="Terminal"}
just check-python
```

::: callout-tip
If you didn't use our template nor do you want to add Ruff to your
project's dependencies, you can use Ruff through
[uv](https://docs.astral.sh/uv/):

``` {.bash filename="Terminal"}
uvx ruff format
# Tries to fix the linted issues, a bit different from `format`
uvx ruff check --fix
uvx ruff check
uvx mypy .
```
:::

> ## Where to write code {#where-to-write-code}

Reviewer 3: Fixed with the addition of "in terms of how to structure your functionality in files and folders, " below.

::: callout-note
Sometimes it's impossible to write in a purely functional style. For
example, when you need to write to a file or database, this action is
inherently a side effect. And that's why the most successful programming
languages, like Python, are those that are flexible in the type of style
you can use to write code. However, Python's functional programming
capabilities are not as strong as those of some other languages, which
can sometimes make it quite difficult to write functional code.
:::

A major benefit to writing more functional code is that it is much
easier to create unit [tests](tests.qmd) for it and it is easier for the
user to understand what the code does.

### Type hints {#type-hints}

For keeping the input and output consistent and clear, it helps to write
[type hints](https://docs.python.org/3/library/typing.html). What are
type hints in Python? They are a way to formally declare the type of
object that is given or output by a function or method. For example, for
the function `check_heart_rate()`, you might write:

``` {.python filename="src/health_data_checks/check_heart_rate.py"}
def check_heart_rate(data: dict[str, Any]) -> bool:

>     """Check if the heart rate is within a normal range."""

Author: Or I'll just remove the docstring :+1:

is predictable and consistent.
:::

### Side effects {#side-effects}

Side effects happen when the code does something that isn't obvious from
the name of the function or method, or when it performs actions without
returning a value.

Some side effects are intentional and can't be avoided, like writing to
a file or database, which are inherently side effects. But in general,
you want to avoid side effects as they introduce unpredictability in
your code.

For example, if you use a function called `check_heart_rate()`, you
probably expect it to output a boolean value that tells you if the heart
rate is normal or not. However, if it also writes to a file or modifies
the data in some way, then it has a side effect. A side effect for
`check_heart_rate()` might look like:

``` {.python filename="src/health_data_checks/check_heart_rate.py"}
import json

def check_heart_rate(data: dict[str, Any]) -> bool:
    """Check if the heart rate is within a normal range."""
    # Side effect: Writing
    with open('file.txt', 'w') as file:
        file.write(json.dumps(data))
    # Pseudocode of a hypothetical check
    not_normal = data["heart_rate"] < 30 or data["heart_rate"] > 180

>     return not_normal

Reviewer 3: Hm, why not return `TRUE` if it's normal? At least, that's what I would expect?

Author: I'm not sure I understand. You mean, use `normal` instead of `not_normal`?

> ```

Reviewer 3: Ah, here my example function with the range by age isn't the best match, maybe. At least, it doesn't fit perfectly with this.

Other side effects include modifying objects in the global scope, such
as modifying the input arguments. Using the same `check_heart_rate()`
function, a side effect that modifies the input might look like this:

``` {.python filename="src/health_data_checks/check_heart_rate.py"}
# Fake data
data = {
    "heart_rate": 30,
}

def check_heart_rate(data: dict) -> bool:
    """Check if the heart rate is within a normal range."""
    # Side effect: Modifying the input in the global scope.
    not_normal = data["heart_rate"] < 30 or data["heart_rate"] > 180
    data.update(check=not_normal)
    return not_normal

# Original data
print(data)
print(check_heart_rate(data))
# Modified data
print(data)
```

A function should never modify objects in the global scope or modify the
input arguments, as it makes it much harder to understand, predict, and
test.

### Internal code

Rarely does a public-facing function or method do everything within
itself. That would end up making the code very long and difficult to
read. Instead, you create smaller, well-defined internal functions,
methods, or classes that do very specific actions and use them to build
up the larger public-facing functionality.

> As already described above in [where to write
> code](#where-to-write-code), when more than one public-facing function
> or method uses the same internal functionality, keep that internal
> functionality in the `internals/` folder.
> 
> How you write this internal functionality is basically no different from
> writing public-facing functionality. The only difference is that you
> should use an underscore (`_`) before the name of the object.

Reviewer 3: ```suggestion
How you write this internal functionality is basically no different from
writing public-facing functionality. The only difference is that the name of internal objects
should start with an underscore (`_`)  and if shared by multiple public-facing functions
or methods, they should go in
the `internals/` folder (as described in [where to write code](#where-to-write-code) above).
```

> Then, once you've made the function or method, in order to "export" it
> to make it easier to import, you open the `__init__.py` file in the
> `internals/` folder and write it like:

Reviewer 3: ```suggestion
To easily export and use the functions in the `internals` module, create an `internals/__init__.py` file
and add the function or method  to it
by writing something like:
```

``` {.python filename="src/health_data_check/internals/__init__.py"}
# Import the `is.py` script by using the relative
# `.` import syntax.
from .is import _is_data_object

# Use `__all__` to explicitly define what is exported
# from this module.
__all__ = [
    "_is_data_object",
]
```

This way, when you import the internal functionality in other scripts,
you can use either:

``` python
from internals import _is_data_object

# Using the function.
_is_data_object(data)
```

Or:

``` python
import internals

# Using the function
internals._is_data_object(data)
```

> We tend to prefer the first option, as explicitly stating the function
> used makes it easier to know what the script is using from internals.

Reviewer 3: ```suggestion
We tend to prefer the first option, as it explicitly states the function
used which makes it easier to know what the script it is from in the `internals/` module.
```

## Writing code documentation

Documentation is one of the most important parts of writing code.
Without good documentation, code is nearly unusable or at least
extremely difficult to use. Documenting at the code level is done with
["docstrings"](https://peps.python.org/pep-0257/) in Python.

At the very least, you should write a docstring for every public-facing

> function, method, and class. This is to help the user understand what

Reviewer 3: ```suggestion
function, method, and class. This helps the user understand what
```

> the code does, what it needs as input, and what it outputs. For internal

Reviewer 3: ```suggestion
the code does, what it needs as input, and what it outputs. You might think this sounds similar to the type hints described above, but docstrings are different: They explain what the code does and expects in natural language, while type checks enforce the correctness of data types. Together, they enhance both code clarity and reliability. For internal
```
Maybe? 
"Since docstrings describe what the code does and expects in words, they are needed alongside type checks."
Docstrings and type checks serve different purposes—while neither is strictly necessary, both are very helpful and complement each other

code, docstrings are not strictly required, but we still recommend
writing them (see guideline below about this).

> When writing docstrings, you should follow these guidelines:

Reviewer 3: Maybe mention [PEP 257](https://peps.python.org/pep-0257/) and somf of the different conventions; e.g., we follow the Google docstring style, but others exist like the NumPy style.

-   Use the `"""` triple quotes for docstrings.
-   Use sentence case for describing things and end with a full stop. So
    use `path: The path to the file.` rather than
    `path: the path to the file`.
-   Include example code in the Quarto code chunk format.
-   Write simple one-line docstrings for:
    -   Very small functions or methods (for example, \<5 lines)
    -   Functions or methods with \<2 arguments
    -   Functions or methods with simple and clear input and output
-   Write longer and more detailed docstrings for:
    -   Larger and/or more complicated internal functions, classes, or
        methods, such as those with non-standard input or output
    -   User-facing functions, classes, or methods

::: callout-tip

> When using VS Code and adding docstrings, use the

Reviewer 3: ```suggestion
When using VS Code and adding docstrings, you can use the
```

[autoDocstring](https://marketplace.visualstudio.com/items?itemName=njpwerner.autodocstring)
VS Code extension to automatically generate a skeleton structure for the
docstrings.
:::

> 

Reviewer 3: ```suggestion

Following the [Google style](https://google.github.io/styleguide/pyguide.html#383-functions-and-methods) for functions or methods, the docstring for the `check_heart_rate()` function looks like this:

```
